diff --git a/core/lib/Drupal/Core/Form/FormBuilder.php b/core/lib/Drupal/Core/Form/FormBuilder.php
index a9f5d8795b..e2bceecbef 100644
--- a/core/lib/Drupal/Core/Form/FormBuilder.php
+++ b/core/lib/Drupal/Core/Form/FormBuilder.php
@@ -578,6 +578,18 @@ public function processForm($form_id, &$form, FormStateInterface &$form_state) {
     $unprocessed_form = $form;
     $form = $this->doBuildForm($form_id, $form, $form_state);
 
+    // Allow an Ajax callback while the form is operating in GET mode. For
+    // example, when using HOOK_form_views_exposed_form_alter.
+    if ($form_state->isMethodType('get')) {
+      $triggering_element_name = $this->requestStack->getCurrentRequest()->request->get('_triggering_element_name');
+      $triggering_element = $form_state->getTriggeringElement();
+      if (isset($triggering_element['#name'])
+        && $triggering_element['#name'] == $triggering_element_name
+        && isset($triggering_element['#ajax'])) {
+        throw new FormAjaxException($form, $form_state);
+      }
+    }
+
     // Only process the input if we have a correct form submission.
     if ($form_state->isProcessingInput()) {
       // Form values for programmed form submissions typically do not include a
diff --git a/core/modules/views/src/ViewExecutable.php b/core/modules/views/src/ViewExecutable.php
index 16baeda36a..e751102643 100644
--- a/core/modules/views/src/ViewExecutable.php
+++ b/core/modules/views/src/ViewExecutable.php
@@ -744,6 +744,11 @@ public function getExposedInput() {
       $this->initDisplay();
 
       $this->exposed_input = $this->request->query->all();
+
+      $request = \Drupal::request();
+      $triggered = $request->request->get('_triggering_element_name');
+      $this->exposed_input = $request->isMethod('post') && $triggered ? $request->request->all() : $request->query->all();
+
       // Unset items that are definitely not our input:
       foreach (['page', 'q'] as $key) {
         if (isset($this->exposed_input[$key])) {
