diff --git a/includes/webform.form_alter.inc b/includes/webform.form_alter.inc
index becbc1daa..9e03ab6a9 100644
--- a/includes/webform.form_alter.inc
+++ b/includes/webform.form_alter.inc
@@ -53,6 +53,7 @@ function _webform_form_webform_submission_form_after_build($form, FormStateInter
  * @see https://www.drupal.org/project/webform/issues/2920095
  */
 function webform_form_update_manager_update_form_alter(&$form, FormStateInterface $form_state) {
+  return;
   if (!isset($form['projects']) || !isset($form['projects']['#options']['webform'])) {
     return;
   }
diff --git a/src/EntitySettings/WebformEntitySettingsFormForm.php b/src/EntitySettings/WebformEntitySettingsFormForm.php
index 6bfaf15c6..30d2fd062 100644
--- a/src/EntitySettings/WebformEntitySettingsFormForm.php
+++ b/src/EntitySettings/WebformEntitySettingsFormForm.php
@@ -77,14 +77,14 @@ class WebformEntitySettingsFormForm extends WebformEntitySettingsBaseForm {
       ':cache_control_override_href' => 'https://www.drupal.org/project/cache_control_override',
     ];
     if ($this->moduleHandler->moduleExists('page_cache') && !$this->moduleHandler->moduleExists('cache_control_override')) {
-      $form['form_settings']['scheduled']['page_cache'] = [
-        '#type' => 'webform_message',
-        '#message_type' => 'warning',
-        '#message_close' => TRUE,
-        '#message_storage' => WebformMessage::STORAGE_SESSION,
-        '#message_message' => $this->t('Scheduled forms do not work as expected for anonymous users when Drupal\'s <a href=":page_cache_href">Internal Page Cache</a> module is enabled. This is a <a href=":issue_href">known issue</a>.', $t_args) . '<br/><br/>' .
-        '<strong>' . $this->t('It is strongly recommended that you install the <a href=":cache_control_override_href">Cache Control Override</a> module.', $t_args) . '</strong>',
-      ];
+      // $form['form_settings']['scheduled']['page_cache'] = [
+      //   '#type' => 'webform_message',
+      //   '#message_type' => 'warning',
+      //   '#message_close' => TRUE,
+      //   '#message_storage' => WebformMessage::STORAGE_SESSION,
+      //   '#message_message' => $this->t('Scheduled forms do not work as expected for anonymous users when Drupal\'s <a href=":page_cache_href">Internal Page Cache</a> module is enabled. This is a <a href=":issue_href">known issue</a>.', $t_args) . '<br/><br/>' .
+      //   '<strong>' . $this->t('It is strongly recommended that you install the <a href=":cache_control_override_href">Cache Control Override</a> module.', $t_args) . '</strong>',
+      // ];
     }
     $form['form_settings']['scheduled']['open'] = [
       '#type' => 'datetime',
