// intercept-client/lib/actions.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.edit = exports.add = exports.setValidating = exports.setTimestamp = exports.markDirty = exports.setSaved = exports.clearErrors = exports.reset = exports.purge = exports.failure = exports.receiveTranslation = exports.receive = exports.request = undefined;

var _actionCreator = require('./util/actionCreator');

var _actionCreator2 = _interopRequireDefault(_actionCreator);

var _actionTypes = require('./actionTypes');

var t = _interopRequireWildcard(_actionTypes);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var request = exports.request = (0, _actionCreator2.default)(t.REQUEST, 'resource', 'id');

var receive = exports.receive = (0, _actionCreator2.default)(t.RECEIVE, 'resp', 'resource', 'id');

var receiveTranslation = exports.receiveTranslation = (0, _actionCreator2.default)(t.RECEIVE_TRANSLATION, 'resp', 'resource', 'langcode', 'id');

var failure = exports.failure = (0, _actionCreator2.default)(t.FAILURE, 'error', 'resource', 'id');

// Removes all local items and resets API syncing state;
var purge = exports.purge = (0, _actionCreator2.default)(t.PURGE, 'resource');

// Resets API syncing state;
var reset = exports.reset = (0, _actionCreator2.default)(t.RESET, 'resource');

var clearErrors = exports.clearErrors = (0, _actionCreator2.default)(t.CLEAR_ERRORS, 'resource', 'id');

var setSaved = exports.setSaved = (0, _actionCreator2.default)(t.SET_SAVED, 'value', 'resource', 'id');

var markDirty = exports.markDirty = (0, _actionCreator2.default)(t.MARK_DIRTY, 'resource', 'id');

var setTimestamp = exports.setTimestamp = (0, _actionCreator2.default)(t.SET_TIMESTAMP, 'resource', 'timestamp');

var setValidating = exports.setValidating = (0, _actionCreator2.default)(t.SET_VALIDATING, 'resource', 'value');

var add = exports.add = (0, _actionCreator2.default)(t.ADD, 'data', 'resource', 'id');

var edit = exports.edit = (0, _actionCreator2.default)(t.EDIT, 'data', 'resource', 'id');
