// intercept-client/lib/constants.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
var NAME = exports.NAME = 'intercept';

// Filters
var DATE = exports.DATE = 'date';
var DATE_START = exports.DATE_START = 'date--start';
var DATE_END = exports.DATE_END = 'date--end';
var KEYWORD = exports.KEYWORD = 'keyword';

// Entity Types
var TYPE_FILE = exports.TYPE_FILE = 'file--file';
var TYPE_SAVED_EVENT = exports.TYPE_SAVED_EVENT = 'flagging--saved_event';
var TYPE_EVENT_RECURRENCE = exports.TYPE_EVENT_RECURRENCE = 'event_recurrence--event_recurrence';
var TYPE_MEDIA_FILE = exports.TYPE_MEDIA_FILE = 'media--file';
var TYPE_MEDIA_IMAGE = exports.TYPE_MEDIA_IMAGE = 'media--image';
var TYPE_MEDIA_SLIDESHOW = exports.TYPE_MEDIA_SLIDESHOW = 'media--slideshow';
var TYPE_MEDIA_VIDEO = exports.TYPE_MEDIA_VIDEO = 'media--web_video';
var TYPE_EQUIPMENT = exports.TYPE_EQUIPMENT = 'node--equipment';
var TYPE_EQUIPMENT_RESERVATION = exports.TYPE_EQUIPMENT_RESERVATION = 'equipment_reservation--equipment_reservation';
var TYPE_EVENT = exports.TYPE_EVENT = 'node--event';
var TYPE_EVENT_ATTENDANCE = exports.TYPE_EVENT_ATTENDANCE = 'event_attendance--event_attendance';
var TYPE_EVENT_SERIES = exports.TYPE_EVENT_SERIES = 'node--event_series';
var TYPE_EVENT_REGISTRATION = exports.TYPE_EVENT_REGISTRATION = 'event_registration--event_registration';
var TYPE_LOCATION = exports.TYPE_LOCATION = 'node--location';
var TYPE_ROOM = exports.TYPE_ROOM = 'node--room';
var TYPE_ROOM_RESERVATION = exports.TYPE_ROOM_RESERVATION = 'room_reservation--room_reservation';
var TYPE_AUDIENCE = exports.TYPE_AUDIENCE = 'taxonomy_term--audience';
var TYPE_EQUIPMENT_TYPE = exports.TYPE_EQUIPMENT_TYPE = 'taxonomy_term--equipment_type';
var TYPE_EVALUATION_CRITERIA = exports.TYPE_EVALUATION_CRITERIA = 'taxonomy_term--evaluation_criteria';
var TYPE_EVENT_TYPE = exports.TYPE_EVENT_TYPE = 'taxonomy_term--event_type';
var TYPE_SUBJECT = exports.TYPE_SUBJECT = 'taxonomy_term--lc_subject';
var TYPE_MEETING_PURPOSE = exports.TYPE_MEETING_PURPOSE = 'taxonomy_term--meeting_purpose';
var TYPE_POPULATION_SEGMENT = exports.TYPE_POPULATION_SEGMENT = 'taxonomy_term--population_segment';
var TYPE_ROOM_TYPE = exports.TYPE_ROOM_TYPE = 'taxonomy_term--room_type';
var TYPE_TAG = exports.TYPE_TAG = 'taxonomy_term--tag';
var TYPE_USER = exports.TYPE_USER = 'user--user';
