// intercept-client/es/actions.js
import actionCreator from './util/actionCreator';
import * as t from './actionTypes';

export var request = actionCreator(t.REQUEST, 'resource', 'id');

export var receive = actionCreator(t.RECEIVE, 'resp', 'resource', 'id');

export var receiveTranslation = actionCreator(t.RECEIVE_TRANSLATION, 'resp', 'resource', 'langcode', 'id');

export var failure = actionCreator(t.FAILURE, 'error', 'resource', 'id');

// Removes all local items and resets API syncing state;
export var purge = actionCreator(t.PURGE, 'resource');

// Resets API syncing state;
export var reset = actionCreator(t.RESET, 'resource');

export var clearErrors = actionCreator(t.CLEAR_ERRORS, 'resource', 'id');

export var setSaved = actionCreator(t.SET_SAVED, 'value', 'resource', 'id');

export var markDirty = actionCreator(t.MARK_DIRTY, 'resource', 'id');

export var setTimestamp = actionCreator(t.SET_TIMESTAMP, 'resource', 'timestamp');

export var setValidating = actionCreator(t.SET_VALIDATING, 'resource', 'value');

export var add = actionCreator(t.ADD, 'data', 'resource', 'id');

export var edit = actionCreator(t.EDIT, 'data', 'resource', 'id');
