// intercept-client/es/constants.js
export var NAME = 'intercept';

// Filters
export var DATE = 'date';
export var DATE_START = 'date--start';
export var DATE_END = 'date--end';
export var KEYWORD = 'keyword';

// Entity Types
export var TYPE_FILE = 'file--file';
export var TYPE_SAVED_EVENT = 'flagging--saved_event';
export var TYPE_EVENT_RECURRENCE = 'event_recurrence--event_recurrence';
export var TYPE_MEDIA_FILE = 'media--file';
export var TYPE_MEDIA_IMAGE = 'media--image';
export var TYPE_MEDIA_SLIDESHOW = 'media--slideshow';
export var TYPE_MEDIA_VIDEO = 'media--web_video';
export var TYPE_EQUIPMENT = 'node--equipment';
export var TYPE_EQUIPMENT_RESERVATION = 'equipment_reservation--equipment_reservation';
export var TYPE_EVENT = 'node--event';
export var TYPE_EVENT_ATTENDANCE = 'event_attendance--event_attendance';
export var TYPE_EVENT_SERIES = 'node--event_series';
export var TYPE_EVENT_REGISTRATION = 'event_registration--event_registration';
export var TYPE_LOCATION = 'node--location';
export var TYPE_ROOM = 'node--room';
export var TYPE_ROOM_RESERVATION = 'room_reservation--room_reservation';
export var TYPE_AUDIENCE = 'taxonomy_term--audience';
export var TYPE_EQUIPMENT_TYPE = 'taxonomy_term--equipment_type';
export var TYPE_EVALUATION_CRITERIA = 'taxonomy_term--evaluation_criteria';
export var TYPE_EVENT_TYPE = 'taxonomy_term--event_type';
export var TYPE_SUBJECT = 'taxonomy_term--lc_subject';
export var TYPE_MEETING_PURPOSE = 'taxonomy_term--meeting_purpose';
export var TYPE_POPULATION_SEGMENT = 'taxonomy_term--population_segment';
export var TYPE_ROOM_TYPE = 'taxonomy_term--room_type';
export var TYPE_TAG = 'taxonomy_term--tag';
export var TYPE_USER = 'user--user';
